File: educational-core/src/com/jetbrains/edu/learning/handlers/EduMoveDelegate.java
package com.jetbrains.edu.learning.handlers;

import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.Messages;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiReference;
import com.intellij.refactoring.move.MoveCallback;
import com.intellij.refactoring.move.MoveHandlerDelegate;
import com.jetbrains.edu.learning.EduUtils;
import org.jetbrains.annotations.Nullable;

import static com.jetbrains.edu.learning.handlers.HandlersUtils.isMoveForbidden;

public class EduMoveDelegate extends MoveHandlerDelegate{
  @Override
  public boolean canMove(DataContext dataContext) {
    return isMoveForbidden(dataContext);
  }

  @Override
  public boolean canMove(PsiElement[] elements, @Nullable PsiElement targetContainer) {
    if (elements.length == 1) {
      Project project = elements[0].getProject();
      return isMoveForbidden(project, elements[0], targetContainer);
    }
    return false;
  }

  @Override
  public boolean isValidTarget(PsiElement psiElement, PsiElement[] sources) {
    return true;
  }

  @Override
  public void doMove(final Project project,
                     PsiElement[] elements,
                     @Nullable PsiElement targetContainer,
                     @Nullable MoveCallback callback) {
    Messages.showInfoMessage("This move operation can break the course", "Invalid Move Operation");
  }

  @Override
  public boolean tryToMove(PsiElement element,
                           Project project,
                           DataContext dataContext,
                           @Nullable PsiReference reference,
                           Editor editor) {
    return EduUtils.isEduProject(project);
  }
}

File: educational-core/src/com/jetbrains/edu/learning/handlers/EduMoveDelegate.kt
package com.jetbrains.edu.learning.handlers

import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiReference
import com.intellij.refactoring.move.MoveCallback
import com.intellij.refactoring.move.MoveHandlerDelegate
import com.jetbrains.edu.learning.EduUtilsKt.isEduProject
import com.jetbrains.edu.learning.messages.EduCoreBundle.message

open class EduMoveDelegate : MoveHandlerDelegate() {
  override fun canMove(dataContext: DataContext): Boolean {
    return isMoveForbidden(dataContext)
  }

  override fun canMove(elements: Array<PsiElement>, targetContainer: PsiElement?, reference: PsiReference?): Boolean {
    return if (elements.size == 1) {
      isMoveForbidden(elements[0].project, elements[0], targetContainer)
    }
    else false
  }

  override fun isValidTarget(psiElement: PsiElement?, sources: Array<PsiElement>): Boolean = true

  override fun doMove(
    project: Project,
    elements: Array<PsiElement>,
    targetContainer: PsiElement?,
    callback: MoveCallback?
  ) {
    Messages.showInfoMessage(message("messages.move.operation.description"), message("messages.move.invalid"))
  }

  override fun tryToMove(
    element: PsiElement,
    project: Project,
    dataContext: DataContext,
    reference: PsiReference?,
    editor: Editor
  ): Boolean {
    return project.isEduProject()
  }
}
